### PR TITLE
SyncTarget Uniqueness

### PR DESCRIPTION
**Pull request by jmprusi**
This PR improves the uniqueness of the labels/annotations/finalizers that rely currently on the SyncTarget Name by replacing it with a hash generated from the SyncTarget Name + SyncTarget workspace.

* Introduces a new helper in `workloadv1alpha1`: `ToSyncTargetKey(syncTargetName, syncTargetWorkspace)`
* Replaces all usages of SyncTarget Name in labels/annotations/finalizers

### CODE REVIEW

File: pkg/syncer/spec/spec_process.go [hunk 1]
```diff
@@ -256,9 +256,10 @@ func (c *Controller) ensureSyncerFinalizer(ctx context.Context, gvr schema.Group
 		namespace := upstreamObjCopy.GetNamespace()
 		logicalCluster := logicalcluster.From(upstreamObjCopy)
 
-		upstreamFinalizers = append(upstreamFinalizers, shared.SyncerFinalizerNamePrefix+c.syncTargetName)
+		upstreamFinalizers = append(upstreamFinalizers, shared.SyncerFinalizerNamePrefix+c.syncTargetKey)
 		upstreamObjCopy.SetFinalizers(upstreamFinalizers)
-		if _, err := c.upstreamClient.Cluster(logicalCluster).Resource(gvr).Namespace(namespace).Update(ctx, upstreamObjCopy, metav1.UpdateOptions{}); err != nil {
+		if _, err := c.upstreamClient.Cluster(logicalCluster).Resource(gvr).Namespace(namespace).Patch(ctx, name, types.MergePatchType, []byte(`{"metadata":{"finalizers":["`+shared.SyncerFinalizerNamePrefix+c.syncTargetKey+`"]}}`), metav1.PatchOptions{}); err != nil {
```
Comment (sttts) [inline pkg/syncer/spec/spec_process.go:261]: this doesn't work I think. A merge patch with a slice will replace the slice. You have to append.

File: pkg/syncer/spec/spec_process.go [hunk 2]
```diff
@@ -349,17 +350,17 @@ func (c *Controller) applyToDownstream(ctx context.Context, gvr schema.GroupVers
...
 	//                as the virtual workspace will set the the deletionTimestamp() on the location view by a transformation.
-	intendedToBeRemovedFromLocation := upstreamObj.GetAnnotations()[workloadv1alpha1.InternalClusterDeletionTimestampAnnotationPrefix+c.syncTargetName] != ""
+	intendedToBeRemovedFromLocation := upstreamObj.GetAnnotations()[workloadv1alpha1.InternalClusterDeletionTimestampAnnotationPrefix+c.syncTargetKey] != ""
 
 	// TODO(jmprusi): When using syncer virtual workspace this condition would not be necessary anymore, since directly tested on the virtual workspace side.
-	stillOwnedByExternalActorForLocation := upstreamObj.GetAnnotations()[workloadv1alpha1.ClusterFinalizerAnnotationPrefix+c.syncTargetName] != ""
-
+	stillOwnedByExternalActorForLocation := upstreamObj.GetAnnotations()[workloadv1alpha1.ClusterFinalizerAnnotationPrefix+c.syncTargetKey] != ""
+	klog.V(4).Infof("Upstream object %s|%s/%s is intended to be removed %b and owned by external actor is: %b", upstreamObjLogicalCluster, upstreamObj.GetNamespace(), upstreamObj.GetName(), intendedToBeRemovedFromLocation, stillOwnedByExternalActorForLocation)
```
Comment (sttts) [inline pkg/syncer/spec/spec_process.go:357]:
```suggestion
	klog.V(4).Infof("Upstream object %s|%s/%s is intended to be removed %b and owned by external actor: %v", upstreamObjLogicalCluster, upstreamObj.GetNamespace(), upstreamObj.GetName(), intendedToBeRemovedFromLocation, stillOwnedByExternalActorForLocation)
```
?

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go [hunk 3]
```diff
@@ -34,7 +34,7 @@ import (
 	"github.com/kcp-dev/kcp/pkg/virtual/framework/internalapis"
 )
 
-func (c *APIReconciler) reconcile(ctx context.Context, apiExport *apisv1alpha1.APIExport, apiDomainKey dynamiccontext.APIDomainKey, syncTargetName string) error {
+func (c *APIReconciler) reconcile(ctx context.Context, apiExport *apisv1alpha1.APIExport, apiDomainKey dynamiccontext.APIDomainKey, syncTargetName string, syncTargetWorkspace logicalcluster.Name) error {
```
Comment (sttts) [inline pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go:37]: unusual order. cluster first, name second.

File: pkg/apis/workload/v1alpha1/helpers.go [hunk 4]
```diff
@@ -17,12 +17,27 @@ limitations under the License.
...
-	value, found := obj.GetLabels()[ClusterResourceStateLabelPrefix+cluster]
+func GetResourceState(obj metav1.Object, syncTargetKey string) (state ResourceState, valid bool) {
+	value, found := obj.GetLabels()[ClusterResourceStateLabelPrefix+syncTargetKey]
 	return ResourceState(value), found && (value == "" || ResourceState(value) == ResourceStateSync)
 }
+
+// ToSyncTargetKey hashes the syncTarget name and workspace to a string that is used to idenfity
+// in a unique way the synctarget in annotations/labels/finalizers.
+func ToSyncTargetKey(syncTargetName string, syncTargetWorkspace logicalcluster.Name) string {
+	hash := sha256.Sum224([]byte(syncTargetName + syncTargetWorkspace.String()))
```
Comment (sttts) [inline pkg/apis/workload/v1alpha1/helpers.go:40]: `syncTargetWorkspace.Join(syncTargetName).String()` please

File: pkg/apis/workload/v1alpha1/helpers.go [hunk 5]
```diff
@@ -17,12 +17,27 @@ limitations under the License.
...
-func GetResourceState(obj metav1.Object, cluster string) (state ResourceState, valid bool) {
-	value, found := obj.GetLabels()[ClusterResourceStateLabelPrefix+cluster]
+func GetResourceState(obj metav1.Object, syncTargetKey string) (state ResourceState, valid bool) {
+	value, found := obj.GetLabels()[ClusterResourceStateLabelPrefix+syncTargetKey]
 	return ResourceState(value), found && (value == "" || ResourceState(value) == ResourceStateSync)
 }
+
+// ToSyncTargetKey hashes the syncTarget name and workspace to a string that is used to idenfity
+// in a unique way the synctarget in annotations/labels/finalizers.
+func ToSyncTargetKey(syncTargetName string, syncTargetWorkspace logicalcluster.Name) string {
```
Comment (sttts) [inline pkg/apis/workload/v1alpha1/helpers.go:39]: switch order, cluster first

File: test/e2e/framework/fixture.go [hunk 6]
```diff
@@ -599,7 +599,7 @@ func (sf SyncerFixture) Start(t *testing.T) *StartedSyncerFixture {
 					// Not a kcp-synced namespace
 					continue
 				}
-				if locator.Workspace.String() != syncerConfig.SyncTargetWorkspace.String() {
+				if locator.SyncTarget.Workspace == syncerConfig.SyncTargetWorkspace.String() {
```
Comment (sttts) [inline test/e2e/framework/fixture.go:602]: why does this only compare the workspace, not the name?

Comment (sttts) [reply test/e2e/framework/fixture.go:602]: and why `!=` => `==` ?

Comment (sttts) [reply test/e2e/framework/fixture.go:602]: https://gist.github.com/a57ad8c7498773cd573d15696e739e1e

File: test/e2e/reconciler/cluster/controller_test.go [hunk 7]
```diff
@@ -67,28 +67,29 @@ func TestClusterController(t *testing.T) {
...
 			work: func(ctx context.Context, t *testing.T, servers map[string]runningServer, syncerFixture *framework.StartedSyncerFixture) {
+				kcpClient, err := kcpclientset.NewForConfig(syncerFixture.SyncerConfig.UpstreamConfig)
+				require.NoError(t, err)
+
+				syncTarget, err := kcpClient.WorkloadV1alpha1().SyncTargets().Get(ctx,
+					syncerFixture.SyncerConfig.SyncTargetName,
+					metav1.GetOptions{},
+				)
+				require.NoError(t, err)
+				syncTargetKey := workloadv1alpha1.ToSyncTargetKey(syncTarget.GetName(), logicalcluster.From(syncTarget))
```
Comment (sttts) [inline test/e2e/reconciler/cluster/controller_test.go:78]: why this change? Just take `syncerFixture.SyncerConfig.SyncTargetName` and `syncerFixture.SyncerConfig.SyncTargetWorkspace` ?

File: test/e2e/reconciler/scheduling/placement_scheduler_test.go [hunk 8]
```diff
@@ -157,6 +157,7 @@ func TestPlacementUpdate(t *testing.T) {
 		},
 	}, metav1.CreateOptions{})
 	require.NoError(t, err)
+	syncTargetKey := workloadv1alpha1.ToSyncTargetKey(firstSyncTargetName, syncerFixture.SyncerConfig.SyncTargetWorkspace)
```
Comment (sttts) [inline test/e2e/reconciler/scheduling/placement_scheduler_test.go:160]: if there is a second synctarget around, call this firstSyncTargetKey

File: test/e2e/syncer/syncer_test.go [hunk 9]
```diff
@@ -311,28 +306,19 @@ func TestSyncerLifecycle(t *testing.T) {
...
-	require.Never(t, func() bool {
+	require.Neverf(t, func() bool {
 		_, err := downstreamKubeClient.AppsV1().Deployments(downstreamNamespaceName).Get(ctx, upstreamDeployment.Name, metav1.GetOptions{})
 		if apierrors.IsNotFound(err) {
 			return true
 		}
 		require.NoError(t, err)
 		return false
-	}, 5*time.Second, time.Second, "downstream Deployment got deleted or there was an error", downstreamNamespaceName, upstreamDeployment.Name)
+	}, 5*time.Second, time.Second, "downstream Deployment %s/%s got deleted or there was an error", downstreamNamespaceName, upstreamDeployment.Name)
```
Comment (sttts) [inline test/e2e/syncer/syncer_test.go:316]:
```suggestion
	}, 5*time.Second, time.Second, "downstream Deployment %s|%s got deleted or there was an error", downstreamNamespaceName, upstreamDeployment.Name)
```

File: test/e2e/reconciler/cluster/controller_test.go (hunk 7 above)
Comment (jmprusi) [reply test/e2e/reconciler/cluster/controller_test.go:77]: I get the syncTarget for the UID (line 91) and then reuse the name/workspace here, but I can use the ones from fixture.

File: pkg/apis/workload/v1alpha1/helpers.go (hunk 5 above)
Comment (jmprusi) [reply pkg/apis/workload/v1alpha1/helpers.go]: done

**Comment by sttts** [comment]
Squashed.

/lgtm
/approve

File: test/e2e/reconciler/scheduling/placement_scheduler_test.go (hunk 8 above)
Comment (jmprusi) [reply test/e2e/reconciler/scheduling/placement_scheduler_test.go]: done
